lib/markdown.js

```diff
@@ -1,35 +1,66 @@
 /* jshint node:true */
 "use strict";
 
+var balanced = require('balanced-match');
+
+var rHeader = /^\s*(\#+|={3,}|-{3,})/;
+var rBadgePrefix = /\[!\[[^\]]*\]\([^)]*\)\]\(/g; // ends with the first parenthesis of link
+
 function markdownBadger(content, imgUrl, linkUrl, altText) {
   var badgeContent = getBadgeMarkdown(imgUrl, linkUrl, altText);
 
   var lines = content.split(/\n/);
-  var idealLine = findIdealLineForInsert(lines) || 0;
+  var idealLine = findLastBadgeLine(lines);
+  if(idealLine >= 0) {
+    var line = lines[idealLine];
+    // count badges and find the prefix end index of the last badge in this line
+    var count = 0;
+    var end;
+    line.replace(rBadgePrefix, function(match, offset) {
+      count++;
+      end = offset + match.length - 1;
+      return match;
+    });
 
-  lines.splice(idealLine, 0, badgeContent);
+    var lastPart = line.substring(end);
+    var result = balanced('(', ')', lastPart);
+    if (!result || count === 1 && result.end === lastPart.length - 1) {
+      // only 1 badge in this line and ends with a badge, or exceptions
+      lines.splice(idealLine + 1, 0, badgeContent);
+    } else {
+      // insert into the same line
+      lines[idealLine] = line.substring(0, end + result.end + 1) + ' ' + badgeContent
+          + lastPart.substring(result.end + 1);
+    }
```

Comment:
Why don't we always just insert into the same line?

Reply:
If existed badges are written in different lines, shall we keep the style?
```
[![Join the chat 1](https://badges.gitter.im/Join%20Chat.svg)](https://en.wikipedia.org/wiki/Saw_(disambiguation))
[![Join the chat 1](https://badges.gitter.im/Join%20Chat.svg)](https://en.wikipedia.org/wiki/Saw_(disambiguation))
// add here seems better, maybe I should update markdown-hasbadge-before/after.md
```

Thanks for your #45. Normally, maintainers like you should have permissions to modify PR branches directly. Well, after confirmed this question, I will cherry-pick it.

Reply:
I think we should insert on the same line for now.

In the future, we could detect the type. I sampled a bunch of various projects

Single line,

 - [`facebook/react`](https://github.com/facebook/react/blame/b05e67e36a7c75f63aa47e716004f4d21f1d0bdc/README.md#L1)
 - [`gulpjs/gulp`](https://github.com/gulpjs/gulp/blame/b065a132a68ee8d68e94a5c650169e45468a3a78/README.md#L8)
 - [`remy/nodemon`](https://github.com/remy/nodemon/blame/6925494fbdf82edb905e72c8712d91e3652e82b1/README.md#L12) (actually uses both single and separate lines)
 - [`SeleniumHQ/selenium`](https://github.com/SeleniumHQ/selenium/blame/749dce430a77c8cd5e1dcc894760f154493eaab6/README.md#L1)

Separate lines,

 - [`expressjs/express`](https://github.com/expressjs/express/blame/3ed5090ca91f6a387e66370d57ead94d886275e1/Readme.md#L5)
 - [`reactjs/redux/`](https://github.com/reactjs/redux/blame/9478a8d4d7649a0dcdea82247b78a7fb57e5e679/README.md#L11)

HTML in markdown,

 - [`babel/babel`](https://github.com/babel/babel/blame/5166eef103fd18bbb3f498cbf7065866d9c1ea83/README.md#L11)
 - [`vuejs/vue`](https://github.com/vuejs/vue/blame/8cc7eaeadd00529e687920f9b837b01a9e96de54/README.md#L3)

Reply:
In order for me to modify this branch directly, you have to check a box on the side, https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/.

Reply:
OK. Let's insert into the same line. In fact, I write badges in the same line, too.

I never changed the checkbox. It is always checked. Wired but never mind, I have done the cherry-pick and fixed a typo. Thanks for your review again.